src/core/qgsvectorlayer.cpp

```diff
@@ -2750,6 +2752,11 @@ long QgsVectorLayer::featureCount() const
          ( mEditBuffer ? mEditBuffer->mAddedFeatures.size() - mEditBuffer->mDeletedFeatureIds.size() : 0 );
 }
 
+bool QgsVectorLayer::empty() const
+{
+  return mDataProvider->empty() && ( !mEditBuffer || mEditBuffer->addedFeatures().empty() );
```

Comment:
Edge case: what about a feature which is added and then deleted in the buffer?

Reply:
Give me a performant approach that implements this and I'll do it for ya.

Reply:
Left as an exercise for the reader ;)